### ISSUE TITLE
Annoying warnings.warn(msg, SourceChangeWarning) when running segment_volume()

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/ivado-medical-imaging#225]
Here is what i get when calling `ivadomed.utils.segment_volume()` from SCT:

~~~
julien-macbook:~/sct_testing_data/t2 $ sct_deepseg -i t2.nii.gz -m t2star_sc

--
Spinal Cord Toolbox (git-jca/2626-deepseg-57b99e394347ad4383ecbd7221234b6a6c9f3138)


Loaded 55 axial slices..
/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Unet' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'torch.nn.modules.container.ModuleList' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'torch.nn.modules.conv.Conv2d' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.Decoder' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
/Users/julien/code/sct/python/envs/venv_sct/lib/python3.6/site-packages/torch/serialization.py:657: SourceChangeWarning: source code of class 'ivadomed.models.UpConv' has changed. you can retrieve the original source code by accessing the object's source attribute or set `torch.nn.Module.dump_patches = True` and use the patch tool to revert the changes.
  warnings.warn(msg, SourceChangeWarning)
../torch/csrc/utils/tensor_numpy.cpp:141: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program.

Done! To view results, type:
fsleyes t2.nii.gz -cm greyscale t2_seg.nii.gz -cm red -a 70.0 &
~~~

**Comment by andreanne-lemay** [comment]
As mentioned in the warning, there are two ways suggested by PyTorch to resolve the warnings:

**Option 1: `set torch.nn.Module.dump_patches = True and use the patch tool to revert the changes.`**

This first solution is a 2 steps process. First thing is to add the suggested line `torch.nn.Module.dump_patches = True` before loading the model. The .patch files will be loaded into the repo and the following warnings will appear: 

``` SourceChangeWarning: source code of class 'ivadomed.models.Unet' has changed. Saved a reverse patch to Unet.patch. Run `patch -p0 < Unet.patch` to revert your changes.```


Afterwards, the patch needs to be reverted to old code using the "patch tool" which is accessible through the command line. This will change the model code to the old one. This solution is not the way to go since it requires manual action + will revert the code.

**Option 2: `retrieve the original source code by accessing the object's source attribute.`**
The second option suggested by PyTorch is:  `you can retrieve the original source code by accessing the object's source attribute`. This is the recommended way to save and load PyTorch models (https://pytorch.org/docs/stable/notes/serialization.html). This implies to save the state_dict instead of the model, so only save the model's parameters. This works well and removes the warnings. The downside is that it requires to have access to the model's code since the code suggested is the following:

`torch.save(the_model.state_dict(), PATH)
the_model = TheModelClass(*args, **kwargs)
the_model.load_state_dict(torch.load(PATH))`

In our case, `TheModelClass` can, for now, be UNet or UNet3D, but we might have more architectures in the future. This will require code addition in `segment_volume` (e.i. `if context['unet_3D']: model = UNet3D() else UNet() ) .

**Option 3: torch.jit**
A third option mentioned by @zougloub is to use torch.jit.save and torch.jit.load (https://pytorch.org/docs/stable/jit.html#torch.jit.ScriptModule). According to the documentation "TorchScript is a way to create serializable and optimizable models from PyTorch code. Any TorchScript program can be saved from a Python process and loaded in a process where there is no Python dependency."  In our model's current form, we can't directly save our Unet() module as TorchScript since we don't only use features supported by TorchScript. I get the following error:

`Type '<class 'ivadomed.models.Unet'>' cannot be compiled`

We would need to `trace` (explained in the documentation I joined above) some modules to make this work which implies modifying the model's code. Since TorchScript is useful to run Pytorch models "high-performance environment such as C++", I don't know if it is the optimal solution to simply get rid of our warning.

**Suggestion**
With all this in mind, my suggestion would be to load and save models the way PyTorch recommends it (option 2). If calling the model's code in segment_volume is not something we want (the downside I described in option 2), I would simply retrain the model with the new code or tolerate the warning. I believe retraining models every time we change the model's code is not viable on the long run, but it could patch this issue for now. Also, if we decide to go with ONNX model it might solve the problem as well.

**Comment by jcohenadad** [comment]
Thanks a lot for your investigations @andreanne-lemay! 🥇 

i would go for option 2, as i don't see the downside as a serious problem for now. And as you said, it is possible that the switch to onnx (which we will investigate in the near future) will solve it.

i would not go for option 1 (re-train model), because if we do another change to the code, the same warning will pop-up again and we are back to zero. This does not sound like a sustainable strategy.

option 3 sounds attractive but i don't know enough the possible implications. Maybe worth further investigating, while also looking at the onnx approach.
my understanding is that if we

**Comment by olix86** [comment]
@andreanne-lemay fixed by #234 ?

**Comment by andreanne-lemay** [comment]
> @andreanne-lemay fixed by #234 ?

Yes! Even if the problem is solved for segment_volume (or at least the models will be converted to .onnx), do we want to implement the option 2 as well (since it's recommended by PyTorch)?

**Comment by andreanne-lemay** [comment]
For now let's keep the current method to save and load models which is saving the whole model, instead of only the state_dict. With ONNX models, the warnings disappear, so we can close this issue.
